src/domain/datasets/services/src/services/statistics/dataset_statistics_indexer.rs

```diff
@@ -0,0 +1,117 @@
+// by the Apache License, Version 2.0.
+
+use std::sync::Arc;
+
+use dill::*;
+use init_on_startup::{InitOnStartup, InitOnStartupMeta};
+use internal_error::{InternalError, ResultIntoInternal};
+use kamu_datasets::DatasetStatisticsRepository;
+
+use crate::{
+    compute_dataset_statistics_increment,
+    JOB_KAMU_DATASETS_DATASET_ENTRY_INDEXER,
+    JOB_KAMU_DATASETS_DATASET_STATISTICS_INDEXER,
+};
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+pub struct DatasetStatisticsIndexer {
+    dataset_storage_unit: Arc<dyn odf::DatasetStorageUnit>,
+    dataset_statistics_repo: Arc<dyn DatasetStatisticsRepository>,
+}
+
+#[component(pub)]
+#[interface(dyn InitOnStartup)]
+#[meta(InitOnStartupMeta {
+    job_name: JOB_KAMU_DATASETS_DATASET_STATISTICS_INDEXER,
+    depends_on: &[
+        JOB_KAMU_DATASETS_DATASET_ENTRY_INDEXER,
+    ],
+    requires_transaction: true,
+})]
+impl DatasetStatisticsIndexer {
+    pub fn new(
+        dataset_storage_unit: Arc<dyn odf::DatasetStorageUnit>,
+        dataset_statistics_repo: Arc<dyn DatasetStatisticsRepository>,
+    ) -> Self {
+        Self {
+            dataset_storage_unit,
+            dataset_statistics_repo,
+        }
+    }
+
+    async fn has_any_stats(&self) -> Result<bool, InternalError> {
+        self.dataset_statistics_repo.has_any_stats().await
+    }
+
+    #[tracing::instrument(
+        level = "debug",
+        skip_all,
+        name = "DatasetStatisticsIndexer::index_dataset_statistics"
+    )]
+    async fn index_dataset_statistics(&self) -> Result<(), InternalError> {
+        use futures::TryStreamExt;
+
+        let dataset_ids: Vec<_> = self
+            .dataset_storage_unit
+            .stored_dataset_ids()
+            .try_collect()
+            .await?;
+
+        for dataset_id in dataset_ids {
+            let dataset = self
+                .dataset_storage_unit
+                .get_stored_dataset_by_id(&dataset_id)
+                .await
+                .int_err()?;
+
+            let head = dataset
+                .as_metadata_chain()
+                .as_uncached_ref_repo()
+                .get(odf::BlockRef::Head.as_str())
+                .await
+                .int_err()?;
+
+            let increment =
+                compute_dataset_statistics_increment(dataset.as_metadata_chain(), &head, None)
+                    .await?;
+
+            assert!(increment.seen_seed);
+            self.dataset_statistics_repo
+                .set_dataset_statistics(&dataset_id, &odf::BlockRef::Head, increment.statistics)
+                .await
+                .int_err()?;
+        }
+
+        Ok(())
+    }
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+#[async_trait::async_trait]
+impl InitOnStartup for DatasetStatisticsIndexer {
+    #[tracing::instrument(
+        level = "debug",
```

Comment:
Indexing seems important (potentially slow) enough to log at `info`. Even better might be for the code that runs all indexers to take care of the tracing in one place.

Reply:
Agree, increased severity in all indexing jobs